tasks/src/test/scala/dagr/tasks/ScatterGatherTests.scala

```diff
@@ -111,4 +115,85 @@ class ScatterGatherTests extends UnitSpec with LazyLogging with BeforeAndAfterAl
     sum1 shouldBe lengths.sum
     sum2 shouldBe lengths.map(x => x*x).sum
   }
+
+  private case class Sample(name: String, library: String, lane: Option[Int])
+
+  private trait SampleTask extends SimpleInJvmTask {
+    var _sample: Option[Sample] = None
+    def sample: Sample = _sample.get
+    def sample_=(s: Sample): Unit = _sample = Some(s)
+  }
+
+  private case class ToSample(input: Path) extends SampleTask {
+    def run(): Unit = {
+      val lines = Io.readLines(input).toSeq
+      require(lines.length == 1, s"Could not parse input: '$input'")
+      sample = lines.head.split(' ') match {
+        case Array(sampleName, library, lane) => Sample(sampleName, library, Some(lane.toInt))
+        case _ => throw new IllegalArgumentException(s"Could not parse line: '${lines.head}'")
+      }
+    }
+  }
+
+  private case class MergeLibraries(libraries: Seq[Sample]) extends SampleTask {
+    def run(): Unit = {
+      require(libraries.map(l => (l.name, l.library)).distinct.length == 1) // same name+lib
+      require(libraries.flatMap(_.lane).distinct.length == libraries.length) // different lanes
+      val library = libraries.map(_.library).mkString(":")
+      sample = libraries.head.copy(library=library, lane=None)
+    }
+  }
+
+  private case class CountLibraries(samples: Seq[Sample], output: Path) extends SimpleInJvmTask  {
+    def run(): Unit = Io.writeLines(output, Seq(samples.map(s => (s.name, s.library)).distinct.length.toString))
+  }
+
+  private case class CountSamples(samples: Seq[Sample], output: Path) extends SimpleInJvmTask  {
+    def run(): Unit = Io.writeLines(output, Seq(samples.map(_.name).distinct.length.toString))
+  }
+
+  it should "run a scatter-gather pipeline with a group-by" in {
+    val lines = Seq("1_A 2_A 1", "1_A 2_A 2", "1_A 2_B 1", "1_B 2_A 1", "1_B 2_B 1")
```

Comment:
Could you please add a comment here on what the columns are on each line?  It would be easier than re-figuring it on re-reading.  And/or change the names to be `s1` for samples and `s1.l1` for libraries?